### PR TITLE
JAC Cloud Orchestrator

### PR DESCRIPTION
**Pull request by AshishMahendra**
## **Description**
JAC Cloud Orchestrator is designed to dynamically import any Python module, deploy it as a Kubernetes Pod, and expose it as an independent gRPC service. This enables any Python module to be a microservice, providing flexibility and scalability in a cloud environment.

### CODE REVIEW

**Comment by AshishMahendra** [comment]
Hi, @marsninja, can you please share your thoughts on the approach, i have added a readme for reference

**Comment by marsninja** [comment]
Good start @AshishMahendra, Some of this implementation looks relevant to the final approach. 

For the first version of jac-cloud-orc (which by the way im thinking if jac-splice-orc might be better) would be the following: 

- A plugin that overrides jac_import to add capability that will automatically turn particular named modules into remote containerized kubernetes pod services that the importing modules would send all interactions to
- Initially these modules can be specified in a config file or through the settings.py interface
- This plugin would use the kubernetes python client library to automatically detect whether these running jac application is in a kubernetes environment and has the right permissions to create pods in that environment (this was prototyped in jaseci 1 and is do-able)
- The plugin would automatically generate the relevant manifests through the k8s protocol to start the pod and initially stall until the pod was ready, then the rest of the module can then execute. 

Basically it would be something like that, i'd want as much automatic control being done by splice-orc as possible.

**Comment by marsninja** [comment]
Hi @AshishMahendra I see you've assigned this to me. What feedback do you want?

**Comment by AshishMahendra** [comment]
Hi @marsninja, we have a version ready, as per your suggestions, so wanted your feedback

File: jac-splice-orc/MANIFEST.in [hunk 1]
```diff
@@ -0,0 +1 @@
+include managers/pod_manager_deployment.yml
```
Comment (marsninja) [inline jac-splice-orc/MANIFEST.in]: Dont think we need this.

File: jac-splice-orc/jac_splice_orc/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,37 @@
+# Use a single base image
```
Comment (marsninja) [inline jac-splice-orc/jac_splice_orc/Dockerfile]: Add some description how its used.

File: jac-splice-orc/jac_splice_orc/managers/pod_manager.py [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+            if result_ip.stdout:
+                return result_ip.stdout.strip()
+
+            if result_hostname.stdout:
+                return result_hostname.stdout.strip()
+
+            return "No LoadBalancer URL found (external IP or hostname missing)."
+
+        except subprocess.CalledProcessError as e:
+            return f"Error executing kubectl command: {e}"
```
Comment (marsninja) [inline jac-splice-orc/jac_splice_orc/managers/pod_manager.py]: Seems like this needs to be removed.

File: jac-splice-orc/jac_splice_orc/managers/pod_manager.py [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+
+    def create_pod(self, module_name: str, module_config: dict) -> Any:
+        """Create a pod and service for the given module."""
+        image_name = os.getenv("IMAGE_NAME", "ashishmahendra/jac-splice-orc:0.1.8")
+
+        pod_name = f"{module_name}-pod"
+        service_name = f"{module_name}-service"
+        print(
+            f"Creating pod {pod_name}...\n Image: {image_name}\n Service: {service_name}\n namespace: {self.namespace}"
+        )
```
Comment (marsninja) [inline jac-splice-orc/jac_splice_orc/managers/pod_manager.py]: prints need to be logger across hte board

File: jac-splice-orc/jac_splice_orc/managers/pod_manager_deployment.yml [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+        app: pod-manager
+    spec:
+      serviceAccountName: smartimportsa  # Use the service account
+      containers:
+      - name: pod-manager
+        image: ashishmahendra/jac-splice-orc:0.1.8
+        ports:
+        - containerPort: 8000
+        env:
+        - name: SERVICE_TYPE
```
Comment (marsninja) [inline jac-splice-orc/jac_splice_orc/managers/pod_manager_deployment.yml]: Dictionaryize, and make naming and selections as config file based.

File: jac-splice-orc/jac_splice_orc/plugin/splice_plugin.py [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+    """JAC Splice-Orchestrator Plugin."""
+
+    def __init__(self):
+        """Constructor for SpliceOrcPlugin."""
+        logging.info("Initializing SpliceOrcPlugin")
+        namespace = "jac-splice-orc"
+        self.create_namespace(namespace)
+        self.create_service_account(namespace)
+        self.apply_pod_manager_yaml(namespace)
+        self.configure_pod_manager_url(namespace)
```
Comment (marsninja) [inline jac-splice-orc/jac_splice_orc/plugin/splice_plugin.py]: Make this not a constructor but just a class method that you call directly, and have it be called by a cli command instead of the deploy.py

File: jac/jaclang/settings.py [hunk 7]
```diff
@@ -27,6 +28,42 @@ class Settings:
...
+            "transformers": {
+                "lib_mem_size_req": "2000Mi",
+                "dependency": ["torch", "transformers"],
+                "lib_cpu_req": "1.0",
+                "load_type": "remote",
+            },
+            "ollama": {
+                "lib_mem_size_req": "300Mi",
+                "dependency": ["ollama"],
+                "lib_cpu_req": "500m",
```
Comment (marsninja) [inline jac/jaclang/settings.py]: there should be no changes to this file.

**Review by marsninja — CHANGES_REQUESTED** [review]

File: jac-splice-orc/MANIFEST.in (hunk 1 above)
Comment (AshishMahendra) [reply jac-splice-orc/MANIFEST.in]: removed

File: jac-splice-orc/jac_splice_orc/managers/pod_manager.py (hunk 3 above)
Comment (AshishMahendra) [reply jac-splice-orc/jac_splice_orc/managers/pod_manager.py]: removed

File: jac-splice-orc/jac_splice_orc/plugin/splice_plugin.py (hunk 6 above)
Comment (AshishMahendra) [reply jac-splice-orc/jac_splice_orc/plugin/splice_plugin.py]: Done, created a command call orc_initialize

File: jac/jaclang/settings.py (hunk 7 above)
Comment (AshishMahendra) [reply jac/jaclang/settings.py]: Removed the changes

File: jac-splice-orc/jac_splice_orc/managers/pod_manager.py (hunk 4 above)
Comment (AshishMahendra) [reply jac-splice-orc/jac_splice_orc/managers/pod_manager.py]: changed

File: jac-splice-orc/jac_splice_orc/Dockerfile (hunk 2 above)
Comment (AshishMahendra) [reply jac-splice-orc/jac_splice_orc/Dockerfile]: Added

File: jac-splice-orc/jac_splice_orc/managers/pod_manager_deployment.yml (hunk 5 above)
Comment (AshishMahendra) [reply jac-splice-orc/jac_splice_orc/managers/pod_manager_deployment.yml]: completed

**Review by marsninja — APPROVED** [review]
